### PR TITLE
Add node selection strategies

### PR DESCRIPTION
**Pull request by sfackler**
## Before this PR
We used the old-school conjure-java-runtime round robin as our only node selection strategy.

## After this PR
==COMMIT_MSG==
The node selection strategy is now configurable and defaults to `PIN_UNTIL_ERROR` to match Dialogue.
==COMMIT_MSG==

The strategies should behave exactly like Dialogue's to the best of my knowledge. This doesn't yet implement updating the strategy based on the server's selection via the response header.

### CODE REVIEW

**Comment by sfackler** [comment]
I've added the balanced strategy as well.

File: conjure-runtime/src/builder.rs [hunk 1]
```diff
@@ -433,3 +430,23 @@ pub enum Idempotency {
...
+    /// periodically to help spread load across the cluster.
+    ///
+    /// This is the default behavior.
+    PinUntilError,
+
+    /// Like `PinUntilError` except that the pinned node is never randomly shuffled.
+    PinUntilErrorWithoutReshuffle,
+
+    /// For each new request, select the "next" node (in some unspecified order).
+    Balanced,
```
Comment (sfackler) [inline conjure-runtime/src/builder.rs:451]: This is called "ROUND_ROBIN" in the c-j-r service config API, but Dialogue uses the term Balanced, including in the server strategy hinting response header so I chose that here.

File: conjure-runtime/src/service/node/selector/single.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    node: Arc<Node>,
+}
+
+impl<S, B> Service<Request<B>> for SingleNodeSelectorService<S>
+where
+    S: Service<Request<B>>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = SingleNodeSelectorFuture<S::Future>;
```
Comment (dtobin) [inline conjure-runtime/src/service/node/selector/single.rs:57]: Is there an advantage to using `SingleNodeSelectorFuture<S::Future>` over `S::Future` here?

[the remark this answers is not in the document]
Comment (sfackler) [reply conjure-runtime/src/service/node/selector/single.rs:57]: The main reason is just to make the enum up in the parent module easier to define if we don't need a special case for single or zero node selectors.

[the remark this answers is not in the document]
Comment (dtobin) [reply conjure-runtime/src/service/node/selector/single.rs:57]: You might be answering a slightly different question than I asked, but I noticed that defining a type here makes `NodeSelectorFuture` cleaner.

[the remark this answers is not in the document]
Comment (sfackler) [reply conjure-runtime/src/service/node/selector/single.rs:57]: Yeah, the NodeSelectorFuture is the enum I was talking about.

[the remark this answers is not in the document]
Comment (dtobin) [reply conjure-runtime/src/service/node/selector/single.rs:57]: Okay cool

**Review by dtobin — APPROVED** [review]
lgtm. It's nice to be able to leverage all of the testing Dialogue is doing on node selection!

**Pull request merged**